### DISCUSSION TITLE
Result of `Object.keys(appRouter._def.procedures)` does not equal it's type

### DISCUSSION THREAD

**Discussion by chrisvander** [trpc/trpc#5521] [Q&A] [answered]
Hi, I'm working on doing permissions on a per-route basis, and I would like to pull out all the procedures in a path oriented format. When I do `console.log(Object.keys(appRouter._def.procedures))`, I get a beautiful list of `['router1.procedure1', 'router1.procedure2', 'router2.procedure1', ...]`.

Unfortunately, the TypeScript type for `keyof appRouter._def.procedures` only returns the top level, aka `'router1' | 'router2'`. How can I pull out a TypeScript type equivalent to what is returned from the `Object.keys` expression?

**Comment by KATT** [comment]
You can't and I don't think we'll support it because it's quite an expensive operation. In v11 we've changed `_.def.procedures` to `Record<string, AnyProcedure>`

You could probably typecast it if you do a helper.

If you want a glance of how I do permissions myself, you can look at https://github.com/trpc/trpc/discussions/5196#discussioncomment-7913438

**Reply by chrisvander** [reply]
So you're explicitly retyping the path of every route as your scope?

**Comment by chrisvander** [comment]
So, based off [the selected StackOverflow answer](https://stackoverflow.com/questions/58434389/typescript-deep-keyof-of-a-nested-object) about deeply nested types, I can walk the `appRouter._def.procedures` with some filtering.

```ts
type Join<K, P> = K extends string | number
  ? P extends "createCaller" | "getErrorShape" | "_def"
    ? never
    : P extends string
      ? `${K}${"" extends P ? "" : "."}${P}`
      : never
  : never
type Prev = [never, 0, 1, 2, 3, 4, 5, ...0[]]
type Leaves<T, D extends number = 4> = [D] extends [never]
  ? never
  : T extends object
    ? {
        [K in keyof T]-?: T[K] extends Record<number | string, AnyProcedure | AnyRouter> | AnyRouter
          ? Join<K, Leaves<T[K], Prev[D]>>
          : K
      }[keyof T]
    : ""

type Procedures = inferRouterDef<AppRouter>["procedures"]
type PathKeys = Leaves<Procedures>
const pathKeys = Object.keys(appRouter._def.procedures) as PathKeys[]
```

Issues I can see: obviously, I'm explicitly cutting off the depth at 4, so if deeper nested routers exist this safety check may need to be removed. I'm also manually filtering out any "createCaller", "getErrorShape", and "_def" named keys, which may not be robust to tRPC updates, but maybe I can pull those out another way.

**Reply by KATT** [reply]
Looks about right. Unsure why it needs to cut at a specific point, should be possible to flatten recursively.

Use `_.def.record` as an entrypoint for this, that's a typed `RouterRecord` that is an easier object:

```ts
export interface RouterRecord {
  [key: string]: AnyProcedure | RouterRecord;
}
```

then you don't need to care about anything else than it being either an object or `extends AnyProcedure` when iterating

**Reply by KATT** [reply, marked as answer]
This was a fun challenge :)
```ts

type ToUnionTuple<
  TRecord extends TRPCRouterRecord,
  TPath extends unknown[] = [],
> = {
  [TKey in keyof TRecord]: TRecord[TKey] extends infer $Value
    ? $Value extends AnyTRPCProcedure
      ? [[...TPath, TKey], $Value]
      : $Value extends TRPCRouterRecord
      ? ToUnionTuple<$Value, [...TPath, TKey]>
      : never
    : never;
}[keyof TRecord];

type Join<TParts, TDelimiter extends string> = TParts extends [
  infer $Head,
  ...infer $Tail,
]
  ? $Head extends string | number
    ? `${$Head}${[] extends $Tail
        ? ''
        : `${TDelimiter}${Join<$Tail, TDelimiter>}`}`
    : never
  : '';

type UnionToIntersection<TUnion> = (
  TUnion extends any ? (x: TUnion) => void : never
) extends (x: infer I) => void
  ? I
  : never;

type Simplify<TType> = { [K in keyof TType]: TType[K] };
type Flatten<TRecord extends TRPCRouterRecord> = Simplify<
  UnionToIntersection<
    ToUnionTuple<TRecord> extends infer $Tuple
      ? $Tuple extends [string[], any]
        ? $Tuple extends [infer $Path, infer $Value]
          ? Record<Join<$Path, '.'>, $Value>
          : never
        : never
      : never
  >
>;

type Flattened = Flatten<typeof appRouter._def.record>;
//    ^?

```

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnmApnA3nAggO0QFQCUAFAYWKggGMUATAVyhQBo4izCJ6YUpCUq0WnAC+cAGaUQcAOQABGFDBUA9AGdeAN14yA3HBUqEyFGoBccHCgAeMAFChIsDHGA5gMdqVESpshUqqGlDaUHoGRkio5pY29o7Q8JgAXj6SENIyyRC0enZ2gjhq8PAAvK7uniSkAHRUTACGPAAUAJS6+YXFcGD0AEYANsBUFNR0jGjlMDVglDQMTB1d8JTcvHBTNas8UB0FEEXwDWBgnGtQG3DbvM3odnBwAOZMKDBujxa9g8Oj8xP3DwibAAFsA1K5wTBgWg3L14GoqNCQA04BBxHBEFwLrMxgsUACHoY4IBeDcAhTtsYBgCz4KCIOCIho4R7vBCg8GM4QaNBRNC8ShQcEHVloKhDFA4eGKBrAR7Ao4AdwaiAJcBqsO4zRVD2SWsBNQgfQAVgIYLddYDLA0QCgLMkasUoO82jUcPQBkM1MC2sxzaJWr6XW6PV7Wj6Lf6LTUAI70XiIZq3Cpwv0bAB8GF1RPwbIhcHlwKaGK4MiYV1ejHcTIQECLjHpYoluqYMArGYtFp4tgsAANoe6awASdDqmAAfhdVrQo9Hsnl0AGuRE3bD7cJRjJFKpbFpcAatFoLJRVnlPUoqFgdOhpY5cC5rngOLADUeTW5NahIobkqg0tlMAAtA0irKquIgdBaIihgCfQQBYdwWn0DQWNM1xQGaq7ArBPT9EMIxzOMTDRrGtIJq0aZtquZYtlAOAUZRCBxBYMiYTIK6geB7aQWxforiIPqQXsPJsBAACq7gHPg9BgAMKAADwAkQAhCHAcTirQ4JeGcOz8IIUC0Cu+DEE0wIqbYangvQOAANY4BA8o4AA2gAupczk+um5TwXADn4AA0igdJuHAVkBWibA6UITnUhFek+f5iAuapODqRU4jrAOABqDQDLGKozpl2WxqZPDJeCuAENUvwEfiq4zg5Dk1I1hnGawfkBU5rAFTlKBObqFhdUVSUpZpXDaUpem6jO+CieJOCSdJckDSw3mNTUzVQq18VOamfWxKEKoWFYoQdCIDkhZi6KKbptBOYJJhwAAUhAbiyc1sBqK1AAiKBDCAHjrEN4IOu8HlsEZ73FeZ3kAm4aUXAOAASKB7iuq2w+l+DSgMPq9Q8+VI3ukOlbeigsgAPpY9AgH0vB5XA3ZDojyO0CIQ7OUTKUDpjwADL6M4yDIvo9kO+Dfb9-1QKz6BPS9XNY19P2gBLqZLkuB17bTDxMXkdhCWJwASRAACSEq8BoVBvAcr36wcoOag8+A27RgO7ngcAzs01jUk7ZGlOmmjPcIh0oPtZEu57FjoxcRu+-7gcAjORsAsHx35EJADKjhDOIiCvfgJig5gDm+RUwWhZd+eoFFbCVygxcuWBuv3QAYgMTQla9MXCC7I3nF3oOZ+A2e5wCTvTSbOzm5bODyRa01j1JMmd+NtDpi7UdwFzi81e2+XzTJHPgg5wNMm5rsJXzm-77yZnEw5G8DkZG2pelWXdbj9Hu3AXeyTLM+Py1WQNQZCpk6m-WMO1P5wBTprSiMCoC7SOrAnaqY7qoDgK3du4o6CXEwTADuPIwrHFOKNXgNQAD6tAUDiC2CvVBdgiSAgAHqjjsEAA

**Reply by chrisvander** [reply]
Very nice! Great workaround indeed. This seems pretty useful for permissions- I wonder if there's some way it makes sense to include in core or document in some other way.

**Reply by chrisvander** [reply]
> 
> 
> Looks about right. Unsure why it needs to cut at a specific point, should be possible to flatten recursively.
> 
> 
> 
> Use `_.def.record` as an entrypoint for this, that's a typed `RouterRecord` that is an easier object:
> 
> 
> 
> ```ts
> 
> export interface RouterRecord {
> 
>   [key: string]: AnyProcedure | RouterRecord;
> 
> }
> 
> ```
> 
> 
> 
> then you don't need to care about anything else than it being either an object or `extends AnyProcedure` when iterating

@KATT I'll mention that while developing it was easier to cut off at a certain depth or else it would hang, so yes, I can probably remove that piece.
